Adds rule "padding-between-test-functions"

tswaters opened the pull request:
Fixes #145 

--

As an aside, I had a heck of a time getting my plugin, as-written, to fit in the eslint ruleset of this project:

- complexity of 4 meant i needed to abuse nested functions a bit... the function really needs complexity of 5 to work properly (exit early if statement; each option has different reports/fixers for padding above/below)

- vars-on-top means I needed to split out the calls to `isPaddingBetweenTokens` into individual functions... it would normally exit early based on results of var declarations, and set up `hasTop/BottomPadding` afterwards.

It's passing the test suite / linting rules for the project, but I fear it has gotten more complex and harder to follow as a result.

lo1tuma commented on lib/rules/padding-between-test-functions.js:
```diff
@@ -0,0 +1,142 @@
+                            message: message,
+                            fix: function (fixer) {
+                                return fixer.replaceTextRange([ lastToken.range[1], tokenAfterLast.range[0] ], '\n');
+                            }
```
Looks like there is a little bit of code duplication in those fixer functions. What do you think about extracting this to a function that creates line removal fixer functions for the given tokens?

Example:

```js
function removeEmptyLinesBetweenTokens(first, second) {
    return function (fixer) {
        return fixer.replaceTextRange([ first.range[1], second.range[0] ], '\n');
    };
}
```

tswaters replied:
Good call, I'll do that.

lo1tuma commented on test/rules/padding-between-test-function.js:
```diff
@@ -0,0 +1,49 @@
+            errors: [ { message: ALWAYS_MESSAGE }, { message: ALWAYS_MESSAGE } ],
+            output: 'describe(function(){\n\nit("", function () {})\n\n})'
+        }
+    ]
```
There is currently also no test case for multiple empty lines between test functions.

tswaters replied:
Hmmm.... yes, multiple empty lines.  The code doesn't handle this case currently.  While it is possible to add an option for, say, number of spaces allowed before/after test functions - I believe this might trample on another eslint rule -- `no-multiple-empty-lines`.  I think it makes the most sense to let the user decide what is best here by setting that rule.

That said, I have noticed a few other problems in adding more tests like you suggested -- the trailing newline doesn't account for semicolons and will dutifully add a lineline between the end of statement `)` and the semicolon.  Working on this now.

lo1tuma replied:
Huh? Isn’t this condition `second.loc.start.line - first.loc.end.line >= 2;` allowing more than one empty line? IMHO this is also the right thing to do, I just wanted to see a test that the rule also works correctly when the padding is more than one empty line.

tswaters commented:
Would you like additional commits or an amended commit with the changed?

lo1tuma commented:
@tswaters Additional commits are ok 😉 .

tswaters commented on lib/rules/padding-between-test-functions.js:
```diff
@@ -0,0 +1,142 @@
+                }
+
+                function always() {
+                    var blockHasTopPadding = isPaddingBetweenTokens(tokenBeforeFirst, firstToken),
```
In my eyes, with never it should be:

```js
[1,2,3].forEach(function () {
  it("foo", function () {});
});
```

...and with always:
```js
[1,2,3].forEach(function () {

  it("foo", function () {});

});
```

The code currently handles this case.  I'll add a test case to ensure this is the case.

lo1tuma replied:
Ok, so in this case I think the rule name might be a bit misleading. The padding is not only enforced **between** (two) test functions but rather **around** every single test function.

tswaters commented on lib/rules/padding-between-test-functions.js:
```diff
@@ -0,0 +1,142 @@
+                function shouldReturn() {
+                    return (
+                        !sourceCode.getTokenBefore(node) ||
+                        !sourceCode.getTokenAfter(node) ||
```
This may be indicative of the complexity of this function - but it'll check both before and after a given mocha function.

So in this case, it checks above describe finds no tokens & aborts; checks below, finds `somethingElse` without padding and reports (if options match)

I've split this out a bit into `topChecker` and `bottomChecker` -- this helps the complexity of the function and hopefully makes things a bit clearer.

I've added test cases to check this particular case.

tswaters commented on lib/rules/padding-between-test-functions.js:
```diff
@@ -0,0 +1,142 @@
+];
+
+/**
+ * Checks if a node is a test function expression, or object (e.g., it() or it.only())
```
Good call, I didn't realize those utility functions existed.  Probably quite a bit cleaner than my `isTestFunction` routine. 

However, because it is split out into 3 different functions, I'm running into complexity problems again due to all the or branches...

I'm going to add the missing hooks, and another util, `isTestFunction` that returns true for any of them.

tswaters commented on docs/rules/padding-between-test-functions.md:
```diff
@@ -0,0 +1,72 @@
+# Padding Between Test Functions (padding-between-test-functions)
+
+This is a stylistic rule that enforces or disallows spaces between test functions.
```
I'll update this a bit to be more clear

tswaters commented:
Hm... found a problem which I think is going to take some time to figure out.  

I was hoping to get this into a stable state today, but it seems this is unlikely.  I'll commit what I have - including some broken test cases - and will come back to it hopefully next weekend.

The problem I'm having is extra errors are being reported (and fixes applied) when there are multiple test blocks that right next to each other.  i.e., for `always`:

```
it()
it()
```

An error is reported for padding after the first one, and padding before the second one, so twice as many errors as should be - and also the fix is being applied for each error so it winds up being:

```
it()


it()
```

lo1tuma commented:
Good catch. I think the best would be to only report the first detected problem and skip the second one. Maybe you could implement this by collecting all possible rule violations first and once you have all then you can filter out the ones with duplicated source code ranges. After that you can report all remaining violations in the `"Program:exit"` listener.

lo1tuma commented on lib/rules/padding-between-test-functions.js:
```diff
@@ -0,0 +1,142 @@
+ * @param {string} message message to use for error reporting.
+ * @returns {function} reporting function that returns array of errors.
+ */
+function getTopCheker(sourceCode, opt, message) {
```
Looks like a typo, I guess it should be getTopChe**c**ker.

tswaters commented:
Just realized I had a typo in that commit message, should be "typo".  leaving it there for ironic purposes.

I've fixed up the problems.  I see ramda is a dependency here, so I tried my best to use it; I've never used it before, so please let me know if there is anything wrong with my usage.

I've added more assertions on the errors to include start/end positions.  The context reporting has also been updated to correctly reflect the start/end line/col where the white space should/should not be.

lo1tuma commented on lib/rules/padding-between-test-functions.js:
```diff
@@ -0,0 +1,166 @@
+                    return;
+                }
+
+                errors = R.reduce(
```
Is the usage of `R.reduce` really necessary here? I think `errors = R.concat(errors, [ topChecker(node), bottomChecker(node) ])` should be sufficient.

tswaters replied:
That's what I tried initially but `R.concat` has arity of 2 - it doesn't care about subsequent parameters.  When passing an array like that it turns it into

```
[
  [
    {top-error},
    {bottom-error},
  ],
  //...etc...
]
```

I stumped across that pattern here: https://stackoverflow.com/a/34376699.

lo1tuma replied:
Ah, you are right, that’s because `topChecker` and `bottomChecker` returns an array as well.

lo1tuma commented on lib/util/ast.js:
```diff
@@ -5,7 +5,8 @@
                         'suite', 'xsuite', 'suite.only', 'suite.skip' ],
-    hooks = [ 'before', 'after', 'beforeEach', 'afterEach' ],
+    hooks = [ 'before', 'after', 'beforeEach', 'afterEach', 'setup', 'tearDown',
+              'suiteSetup', 'suiteTeardown' ],
```
👍

lo1tuma commented on test/rules/padding-between-test-function.js:
```diff
@@ -0,0 +1,258 @@
+            } ],
+            output: [
+                'describe(function(){',
+                'var someVar=null;',
```
Is this really the expected behavior or should there be another blank line between `describe` and `var`?

lo1tuma replied:
I think I got it now. I was confused for a moment and thought the padding of the `describe` block should also affect the inner part, but that doesn’t make sense after a second thought.

lo1tuma commented on docs/rules/padding-between-test-functions.md:
```diff
@@ -0,0 +1,72 @@
+# Padding Between Test Functions (padding-between-test-functions)
```
So what do you think about the rule name? As I understand it the rule enforces the padding around each test function and not specifically between two test functions. So maybe a better name would be `padding-around-test-functions`?

jfmengels commented:
This is a great idea. I was actually thinking of writing this rule myself.

One thing that bothers me though: The way I write my tests is like this, so the same thing as with the `always` option, but without the empty line at the start of `describe`.

Example:
```js
describe('some describe', () => {
  it('shouldnt have spacing', () => {
  });

  it('shouldnt have spacing', () => {
  });

  it('shouldnt have spacing', () => {
  });

  describe('some nested describe', () => {
    it('shouldnt have spacing', () => {
    });
  });
})
```

I have also turned on the core ESLint [padded-blocks](https://eslint.org/docs/rules/padded-blocks) rule, which forbids an empty line at the start of the function. When autofixing, having the two rules would make ESLint go back and forth here.

I would suggest to either:
- Change the rule to enforce the style I just described. If chosen, the rule might be better called `spacing-between-test-functions`
- Add an extra option dealing that supports this style (and also warn users about the potential conflict with `padded-blocks`)

Tell me what you think. And thanks @tswaters for your work on this! :)

tswaters commented:
@jfmengels how about an option for `padded-blocks`  with the same options?  This would inspect any of the `describeAliases` and enforce or forbid padding similar to how `padded-blocks` does it.

AFAIK, eslint rules can't read other config options so a user of this rule would need to specify the same, if desired... otherwise, like you say, eslint would toggle between the two.

I'd need to think about how that would work and I'm not sure I'll have time in the immediate future to work on this.  I've given you both contributor access to my fork if you want a kick at the can 😄

mschilde commented:
Is there any chance that this PR gets merged anytime soon? How can I help?

trollepierre commented:
Hi guys, what's up with this PR ?

pkuczynski commented:
Yeah, I would love to see this PR completed as @jfmengels described. @lo1tuma what's missing so I can maybe help?

tswaters commented:
Hi sorry for not responding here.

If anyone is interested here let me know and I can grant access to my fork.  Another option would be to fork my fork and go from there.  

The changes are all on master, from a while back --so you're probably going to want to rebase that commit against origin.

I don't have any time to work on this so I'm going to  close this PR, and someone else can take up the mantle.
